## CAPM (Capital Asset Pricing Model) to calculate Expected Return of an Asset

- Author: Ayush Patel
- Github: https://github.com/Ayush-Patel15 
- LinkedIn: https://www.linkedin.com/in/ayush-15-patel/

<hr>

As an investor, you may want some assurance that your money will grow and net you a profit. While it practically impossible to predict exactly how much your investments will return, there are methods generally accepted by the investing industry that you can used to estimate future returns based on key assumptions. An expected return is a statistical measures that investors can use to analyze their portfolios or a single asset. The expected return is the anticipated amount of returns that a stock may generate.

### What is CAPM ?

The CAPM formula is used for calculating the expected returns of an asset. It is based on the idea of systematic risk that investors need to be compensated for in the form of a risk premium. It is a model that describes the relationship between the expected return and risk of investing in a security. It shows that the expected return on a security is equal to the risk-free return plus a risk premium, which is based on the beta of that security. A risk premium is a rate of return greater than the risk-free rate. When investing, investors desire a higher risk premium when taking on more risky investments.

The expected return is generally based on `historical returns`. As such, it doesn't indicate the potential for future performance and shouldn't be used as the only decision-making tool. This metric can, however, give investors a reasonable expectation of what they may expect in the short and long-run.

<hr>

### Formula of CAPM

The formula for calculating the expected return of an asset, given its risk, is as follows:

> <div style="margin-left: 40%;">E(R<sub>a</sub>) = R<sub>f</sub> + β<sub>a</sub> * (E(R<sub>m</sub>) - R<sub>f</sub>)</div>

- Expected Return of Asset (E(R<sub>a</sub>)): Expected return of the asset, the formula combines all the above components to calculate the return an investor should expect for taking on the risk of investing in a specific stock.
- Risk-Free Rate (R<sub>f</sub>): This is the return you can get from a completely risk-free investment, like a 10-year government bond.
- Beta (β<sub>a</sub>): Beta measures how much a stock’s price moves in relation to the market.
- Expected Market Return (E(R<sub>m</sub>)): Expected return of the market index
- Market Risk Premium (E(R<sub>m</sub>) - R<sub>f</sub>): This represents the extra return investors expect for taking on market risk over the risk-free rate.

### Example of a Stock ABC

Imagine you’re evaluating Stock ABC:

- Risk-Free Rate (R<sub>f</sub>) = 4%
- Beta (β<sub>a</sub>) = 1.3 (Stock ABC is 30% more volatile than the market)
- Expected Market Return (E(R<sub>m</sub>)) = 12%

Therefore, by using the CAPM formula:

> <div style="margin-left: 40%;">E(R<sub>a</sub>) = 4 + 1.3 * (12 - 4) = 14.4%</div>

This means that based on its risk (beta), Stock ABC should generate a return of 14.4% for it to be considered a worthwhile investment.

<hr>

### Importance of CAPM

- Decision-Making: Helps investors evaluate whether a stock is worth the risk.
- Portfolio Optimization: Aids in determining the mix of high-beta (risky) and low-beta (stable) assets for diversification.
- Cost of Equity: Widely used in corporate finance to calculate the cost of equity for valuation purposes.

### Advantages of CAPM

- Simplicity: It provides a straightforward framework for evaluating the risk-return tradeoff.
- Focus on Systematic Risk: It only considers market risk (beta), which cannot be diversified away.
- Universal Applicability: Can be used for individual assets, portfolios, or entire sectors.

### Disadvantages of CAPM

- Assumptions Are Unrealistic:
	- Assumes a single factor (market risk) drives returns.
	- Assumes investors have access to the same information and behave rationally.
	- Assumes risk-free borrowing and lending, which doesn’t exist in reality.

- Beta Limitations:
	- Beta is not static and can change over time.
	- It only measures historical volatility and may not accurately predict future risk.

- Empirical Limitations: In practice, CAPM often fails to fully explain stock returns, especially for smaller companies or emerging markets.

<hr>

### Python program to calculate Expected Return using CAPM

Let's utilize the power of python and calculate the expected return for all Nifty50 Stocks, using the CAPM formula. 

`List of all stocks is`: ['ADANIENT', 'ADANIPORTS', 'APOLLOHOSP', 'ASIANPAINT', 'AXISBANK', 'BAJAJ-AUTO', 'BAJAJFINSV', 'BAJFINANCE', 'BEL', 'BHARTIARTL', 'BPCL', 'BRITANNIA', 'CIPLA', 'COALINDIA', 'DRREDDY', 'EICHERMOT', 'GRASIM', 'HCLTECH', 'HDFCBANK', 'HDFCLIFE', 'HEROMOTOCO', 'HINDALCO', 'HINDUNILVR', 'ICICIBANK', 'INDUSINDBK', 'INFY', 'ITC', 'JSWSTEEL', 'KOTAKBANK', 'LT', 'M&M', 'MARUTI', 'NESTLEIND', 'NTPC', 'ONGC', 'POWERGRID', 'RELIANCE', 'SBILIFE', 'SBIN', 'SHRIRAMFIN', 'SUNPHARMA', 'TATACONSUM', 'TATAMOTORS', 'TATASTEEL', 'TCS', 'TECHM', 'TITAN', 'TRENT', 'ULTRACEMCO', 'WIPRO']

For the analysis:

- Data included for the analysis is: 2014-01-01 to 2024-12-27
- India's 10Y Goverment Bond Yield as of 2024-12-27 is: 6.780% 
- For calculation of Beta, I have taken the Nifty50 index as the broad market index
- For Expected Return of Market, I have used the Historical Annual Return of Nifty50

In [24]:
## Import Statements
import pandas as pd
import numpy as np
import warnings
import os
warnings.filterwarnings("ignore")

## FILE AND FOLDERPATHS
NIFTY50_FILEPATH = "E:\\Market-Work\\All-Data\\Daily_Data-Stocks_Indices\\NIFTY50.csv"
STOCKS_FOLDERPATH = "E:\\Market-Work\\All-Data\\Daily_Data-Stocks_Indices"
all_stocks_list = ['ADANIENT', 'ADANIPORTS', 'APOLLOHOSP', 'ASIANPAINT', 'AXISBANK', 'BAJAJ-AUTO', 'BAJAJFINSV', 'BAJFINANCE', 'BEL', 'BHARTIARTL', 
    'BRITANNIA', 'CIPLA', 'COALINDIA', 'DRREDDY', 'EICHERMOT', 'GRASIM', 'HCLTECH', 'HDFCBANK', 'HDFCLIFE', 'HEROMOTOCO', 'HINDALCO', 
	'HINDUNILVR', 'ICICIBANK', 'INDUSINDBK', 'INFY', 'ITC', 'JSWSTEEL', 'KOTAKBANK', 'LT', 'M&M', 'MARUTI', 'NESTLEIND', 'NTPC', 'ONGC', 'POWERGRID', 
    'RELIANCE', 'SBILIFE', 'SBIN', 'SHRIRAMFIN', 'SUNPHARMA', 'TATACONSUM', 'TATAMOTORS', 'TATASTEEL', 'TCS', 'TECHM', 'TITAN', 'TRENT', 'ULTRACEMCO', 'WIPRO'
]
risk_free_rate = 0.06780

## Read the Nifty file, and calculate the returns on annual basis
nifty_df = pd.read_csv(NIFTY50_FILEPATH, index_col=0, usecols=["Date","Close"]).reset_index()
nifty_df["Date"] = pd.to_datetime(nifty_df["Date"], format="%Y-%m-%d")
nifty_df.set_index("Date", inplace=True)
## Calculate the log daily returns and multiply it by 252
nifty_df["Returns"] = np.log(nifty_df["Close"] / nifty_df["Close"].shift())
nifty_df.dropna(inplace=True)
expected_market_return = nifty_df["Returns"].mean() * 252
print("Historical Annual Market Return of Nifty50 is:", round(expected_market_return * 100 , 2))

## Final dictionary to store the expected returns of assets
all_expected_returns_dict = {"Instrument_name": ["Nifty50"], "Beta": [1], "Expected_return": [round(expected_market_return * 100 , 2)]}

## Iterate through each stock, and add the answer to the dictionary
for each_stock in all_stocks_list:
	# print("==> Processing:", each_stock)
	stock_filepath = os.path.join(STOCKS_FOLDERPATH, each_stock+".csv")
	stock_df = pd.read_csv(stock_filepath, index_col=0, usecols=["Date","Close"]).reset_index()
	stock_df["Date"] = pd.to_datetime(stock_df["Date"], format="%Y-%m-%d")
	stock_df.set_index("Date", inplace=True)
	stock_df["Returns"] = np.log(stock_df["Close"] / stock_df["Close"].shift())
	stock_df.dropna(inplace=True)

	## Calculate the Beta of the stock
	main_df = pd.merge(left=nifty_df, right=stock_df, left_index=True, right_index=True, suffixes=["_NIFTY", f"_{each_stock}"])
	main_df.dropna(inplace=True)
	covariance = np.cov(main_df[f"Returns_{each_stock}"], main_df["Returns_NIFTY"])[0, 1]
	variance = np.var(main_df["Returns_NIFTY"])
	stock_beta = covariance / variance

	## Calculate the expected return, using the CAPM formula
	expected_stock_return = risk_free_rate + (stock_beta * (expected_market_return - risk_free_rate))
	expected_stock_return_pct = round(expected_stock_return * 100, 2)
	all_expected_returns_dict["Instrument_name"].append(each_stock)
	all_expected_returns_dict["Beta"].append(stock_beta)
	all_expected_returns_dict["Expected_return"].append(expected_stock_return_pct)

## Create a dataframe of the resultant dict
capm_df = pd.DataFrame(all_expected_returns_dict)

Historical Annual Market Return of Nifty50 is: 10.97


In [25]:
## Save it in a csv
capm_df.to_csv("Expected_Returns.csv")
capm_df

Instrument_name      Beta  Expected_return
0          Nifty50  1.000000            10.97
1         ADANIENT  1.060396            11.22
2       ADANIPORTS  1.055775            11.20
3       APOLLOHOSP  0.548515             9.08
4       ASIANPAINT  0.458792             8.70
5         AXISBANK  1.224970            11.91
6       BAJAJ-AUTO  0.726029             9.82
7       BAJAJFINSV  0.859767            10.38
8       BAJFINANCE  0.827886            10.25
9              BEL  0.788481            10.08
10      BHARTIARTL  0.868978            10.42
11       BRITANNIA  0.361919             8.30
12           CIPLA  0.563005             9.14
13       COALINDIA  0.884725            10.49
14         DRREDDY  0.484944             8.81
15       EICHERMOT  0.768302            10.00
16          GRASIM  0.891681            10.52
17         HCLTECH  0.887797            10.50
18        HDFCBANK  0.941373            10.72
19        HDFCLIFE  0.915325            10.62
20      HEROMOTOCO  0.754455             9.94
21        HINDALCO  1.254331            12.04
22      HINDUNILVR  0.585426             9.23
23       ICICIBANK  1.305569            12.25
24      INDUSINDBK  1.279202            12.14
25            INFY  0.779755            10.05
26             ITC  0.679144             9.63
27        JSWSTEEL  1.287435            12.17
28       KOTAKBANK  1.125873            11.50
29              LT  1.116533            11.46
30             M&M  1.019502            11.05
31          MARUTI  0.911704            10.60
32       NESTLEIND  0.428126             8.57
33            NTPC  0.797763            10.12
34            ONGC  1.063448            11.24
35       POWERGRID  0.798807            10.13
36        RELIANCE  1.100325            11.39
37         SBILIFE  0.782253            10.06
38            SBIN  1.177723            11.71
39      SHRIRAMFIN  0.796321            10.12
40       SUNPHARMA  0.534716             9.02
41      TATACONSUM  0.772478            10.02
42      TATAMOTORS  1.197866            11.80
43       TATASTEEL  1.510508            13.11
44             TCS  0.773058            10.02
45           TECHM  1.006205            11.00
46           TITAN  0.922713            10.65
47           TRENT  0.610455             9.34
48      ULTRACEMCO  0.817002            10.20
49           WIPRO  0.880516            10.47

In [26]:
## Minimum and maximum return stocks
min_index = capm_df["Expected_return"].idxmin()
max_index = capm_df["Expected_return"].idxmax()
print("Minimum Expected Return Stock is:", capm_df.loc[min_index, "Instrument_name"], "&& its Expected Return is:", capm_df.loc[min_index, "Expected_return"])
print("Maximum Expected Return Stock is:", capm_df.loc[max_index, "Instrument_name"], "&& its Expected Return is:", capm_df.loc[max_index, "Expected_return"])

Minimum Expected Return Stock is: BRITANNIA && its Expected Return is: 8.3
Maximum Expected Return Stock is: TATASTEEL && its Expected Return is: 13.11


### Learning or Conclusions

- Not suprisingly, the lowest and highest expected returns, are of the lowest and highest beta stocks respectively. The entire calculation of CAPM is `highly dependent on the Beta values` of the stock. Which is not a good case, as an investor or a portfolio manager, one should never rely entirely on a single metric. It's essential to evaluate other metrics like P&L, PE ratio, Sharpe Ratio, and etc. 

- Also appreciate the fact that the expected return of all stocks are somewhat near to the historical returns of Nifty (10.97) which signifies the high correlation of all its components.

- Interestingly, when I calculated the standard deviation of Nifty Returns, it comes out to be (0.0138 or 1.38%). And the Expected returns range is from `10.97 - (2 * sd)` to `10.97 + (2 * sd)` i.e. `8.21` and `13.73` respectively.

- For excercise, try to calculate the expected return of any Midcap and Smallcap stock. You will understand, how beta or volatility of the stock contribute to its expected returns.

<hr>

`DISCLAIMER`: This is just for the purpose of knowledge and my own learning, and not any advice to create any kind of portfolio based on the expected returns values. Do you own analysis and invest in markets. However, this is just for my own learning purpose, so need not to worry. For learning purpose, I can test for any weird possibilities, even if doesn't make any sense in the practical world.

Thank you for reading the complete section. If you have any suggestions or doubts, comment down below or message me on linkedIn.

<hr>